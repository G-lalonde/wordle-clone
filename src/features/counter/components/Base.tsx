import React from "react";
import { StyleSheet, Text, View, Button } from "react-native";
import { useDispatch, useSelector } from "@/app/hooks";
import { add, fetchAsync, remove } from "../actions/counterActions";

const Base = () => {
  const { count } = useSelector(state => state.counter);
  const dispatch = useDispatch();

  const incrementHandler = () => {
    dispatch(add(1));
  };

  const decrementHandler = () => {
    dispatch(remove(1));
  };

  const fetchHandler = () => {
    dispatch(fetchAsync());
  };

  return (
    <View style={styles.container}>
      <View style={styles.counterView}>
        <Text style={styles.count}>Counter</Text>
        <Text style={styles.count}>{count}</Text>
      </View>
      <View style={styles.buttonContainer}>
        <View style={styles.button}>
          <Button
            testID="decrement-button"
            title="-"
            onPress={decrementHandler}
          />
        </View>
        <View style={styles.button}>
          <Button
            testID="increment-button"
            title="+"
            onPress={incrementHandler}
          />
        </View>
      </View>
      <Button testID="increment-button" title="fetch" onPress={fetchHandler} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
  count: {
    fontSize: 20,
  },
  counterView: {
    justifyContent: "center",
    alignItems: "center",
  },
  buttonContainer: {
    flexDirection: "row",
    justifyContent: "space-around",
  },
  button: {
    fontSize: 16,
    width: "20%",
    marginHorizontal: 10,
    marginTop: 20,
  },
});

export default Base;
