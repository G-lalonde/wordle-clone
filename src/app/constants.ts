export const ROW_COUT = 6;
export const ANSWER = "react";
