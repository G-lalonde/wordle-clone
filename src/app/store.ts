import { counterReducer } from "@/features/counter";
import { wordleReducer } from "@/features/wordle";

import {
  Action,
  applyMiddleware,
  combineReducers,
  compose,
  createStore,
  ThunkAction,
} from "@reduxjs/toolkit";
import thunk from "redux-thunk";

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

// PUT ALL REGISTERED REDUCERS HERE
export const rootReducer = combineReducers({
  counter: counterReducer,
  wordle: wordleReducer,
});

export const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk)),
);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
