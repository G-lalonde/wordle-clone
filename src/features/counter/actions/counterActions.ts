import { Dispatch } from "@reduxjs/toolkit";
import {
  AddCounter,
  ADD_COUNTER,
  FetchDispatchType,
  FETCH_FAILED,
  FETCH_REQUEST,
  FETCH_SUCCESS,
  RemoveCounter,
  REMOVE_COUNTER,
} from "./counterActionsTypes";

// EXAMPLE OF SYNC ACTION
export const add = (count: number): AddCounter => ({
  type: ADD_COUNTER,
  payload: count,
});

export const remove = (count: number): RemoveCounter => ({
  type: REMOVE_COUNTER,
  payload: count,
});

// EXAMPLE OF ASYNC ACTION
export const fetchAsync =
  () => async (dispatch: Dispatch<FetchDispatchType>) => {
    dispatch({ type: FETCH_REQUEST });

    await fetchSecretSauce()
      .then(response => response.json())
      .then(
        response => {
          dispatch({
            type: FETCH_SUCCESS,
            payload: response,
          });
        },
        error => dispatch({ type: FETCH_FAILED, payload: error }),
      );
  };

const fetchSecretSauce = () => {
  return fetch(
    "https://raw.githubusercontent.com/oblador/react-native-vector-icons/master/glyphmaps/FontAwesome5Free.json",
  );
};
