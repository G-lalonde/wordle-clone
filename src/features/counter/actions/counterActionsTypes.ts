export const ADD_COUNTER = "ADD_COUNTER";
export const REMOVE_COUNTER = "REMOVE_COUNTER";
export const FETCH_REQUEST = "FETCH_REQUEST";
export const FETCH_SUCCESS = "FETCH_SUCCESS";
export const FETCH_FAILED = "FETCH_FAILED";

export interface AddCounter {
  type: typeof ADD_COUNTER;
  payload: number;
}

export interface RemoveCounter {
  type: typeof REMOVE_COUNTER;
  payload: number;
}

interface FetchRequest {
  type: typeof FETCH_REQUEST;
}

interface FetchRequest {
  type: typeof FETCH_REQUEST;
}

interface FetchSuccess {
  type: typeof FETCH_SUCCESS;
  payload: { [key: string]: string };
}

interface FetchFailed {
  type: typeof FETCH_FAILED;
  payload: string;
}

export type FetchDispatchType = FetchRequest | FetchSuccess | FetchFailed;
