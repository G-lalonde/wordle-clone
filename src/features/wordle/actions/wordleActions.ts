import {
  SetWord1,
  SetWord2,
  SetWord3,
  SetWord4,
  SetWord5,
  SetAnswer,
  SET_WORD1,
  SET_WORD2,
  SET_WORD3,
  SET_WORD4,
  SET_WORD5,
  SET_ANSWER,
} from "./wordleActionsTypes";

export const setAnswer = (word: string): SetAnswer => ({
  type: SET_ANSWER,
  payload: word,
});

export const setWord1 = (word: string): SetWord1 => ({
  type: SET_WORD1,
  payload: word,
});

export const setWord2 = (word: string): SetWord2 => ({
  type: SET_WORD2,
  payload: word,
});

export const setWord3 = (word: string): SetWord3 => ({
  type: SET_WORD3,
  payload: word,
});

export const setWord4 = (word: string): SetWord4 => ({
  type: SET_WORD4,
  payload: word,
});

export const setWord5 = (word: string): SetWord5 => ({
  type: SET_WORD5,
  payload: word,
});
