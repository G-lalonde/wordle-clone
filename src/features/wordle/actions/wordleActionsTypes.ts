export const SET_WORD = "SET_WORD";
export const ADD_CORRECT_LETTER = "ADD_CORRECT_LETTER";
export const ADD_ALMOST_LETTER = "ADD_ALMOST_LETTER";
export const ADD_WRONG_LETTER = "ADD_WRONG_LETTER";
export const SET_ANSWER = "SET_ANSWER";
export const SET_WORD_INDEX = "SET_WORD_INDEX";

export interface SetAnswer {
  type: typeof SET_ANSWER;
  payload: string;
}

export interface SetWord {
  type: typeof SET_WORD;
  payload: string;
}

export interface AddCorrectLetter {
  type: typeof ADD_CORRECT_LETTER;
  payload: string;
}

export interface AddAlmostLetter {
  type: typeof ADD_ALMOST_LETTER;
  payload: string;
}

export interface AddWrongLetter {
  type: typeof ADD_WRONG_LETTER;
  payload: string;
}

export interface SetWordIndex {
  type: typeof SET_WORD_INDEX;
  payload: number;
}

export type WordleActionTypes =
  | SetAnswer
  | SetWord
  | AddCorrectLetter
  | AddAlmostLetter
  | AddWrongLetter
  | SetWordIndex;
