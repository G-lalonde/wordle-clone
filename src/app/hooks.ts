import { FontAwesome } from "@expo/vector-icons";
import * as Font from "expo-font";
import * as SplashScreen from "expo-splash-screen";
import { useEffect, useState } from "react";
import {
  TypedUseSelectorHook,
  useDispatch as useReduxDispatch,
  useSelector as useReduxSelector,
} from "react-redux";
import { AppDispatch, RootState } from "./store";
export const useDispatch = () => useReduxDispatch<AppDispatch>();
export const useSelector: TypedUseSelectorHook<RootState> = useReduxSelector;
import { Mulish_600SemiBold, Mulish_700Bold } from "@expo-google-fonts/mulish";

export function useCachedResources() {
  const [isLoadingComplete, setLoadingComplete] = useState(false);

  // Load any resources or data that we need prior to rendering the app
  useEffect(() => {
    async function loadResourcesAndDataAsync() {
      try {
        SplashScreen.preventAutoHideAsync();

        // Load fonts
        await Font.loadAsync({
          ...FontAwesome.font,
          mulish: require("../../assets/fonts/Mulish-VariableFont_wght.ttf"),
          mulish_semi_bold: Mulish_600SemiBold,
          mulish_bold: Mulish_700Bold,
        });
      } catch (e) {
        // We might want to provide this error information to an error reporting service
        console.warn(e);
      } finally {
        setLoadingComplete(true);
        SplashScreen.hideAsync();
      }
    }

    loadResourcesAndDataAsync();
  }, []);

  return isLoadingComplete;
}
