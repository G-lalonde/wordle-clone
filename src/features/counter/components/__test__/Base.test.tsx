import { store } from "@/app/store";
import { fireEvent, render, shallow } from "@/test-utils";
import React from "react";
import { ADD_COUNTER, REMOVE_COUNTER } from "../../actions/counterActionsTypes";
import Base from "../Base";

// https://testing-library.com/docs/react-native-testing-library/example-intro
// https://jestjs.io/fr/docs/snapshot-testing

const dispatch = jest.spyOn(store, "dispatch");

describe("<Base />", () => {
  it("renders correctly", () => {
    const tree = render(<Base />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("Testing disptached actions", () => {
  it("clic increment disptach ADD_COUNTER action", () => {
    const { getByTestId } = render(<Base />);

    const button = getByTestId("increment-button");
    fireEvent.press(button);

    expect(dispatch).toHaveBeenCalledWith({
      type: ADD_COUNTER,
      payload: 10,
    });
  });

  it("clic decrement disptach REMOVE_COUNTER action", () => {
    const { getByTestId } = render(<Base />);

    const button = getByTestId("decrement-button");
    fireEvent.press(button);

    expect(dispatch).toHaveBeenCalledWith({
      type: REMOVE_COUNTER,
      payload: 2,
    });
  });
});
