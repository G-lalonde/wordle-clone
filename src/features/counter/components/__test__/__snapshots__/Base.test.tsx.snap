// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Base /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Counter
    </Text>
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      0
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <View
      style={
        Object {
          "fontSize": 16,
          "marginHorizontal": 10,
          "marginTop": 20,
          "width": "20%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="decrement-button"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            -
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "fontSize": 16,
          "marginHorizontal": 10,
          "marginTop": 20,
          "width": "20%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="increment-button"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            +
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
